File: codes/go/gemini/minimal/task_2/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"sort"
)

func reallocateBed(patient map[string]interface{}, hospitals []interface{}) map[string]interface{} {
	patientID := patient["id"].(string)
	patientGender := patient["gender"].(string)
	patientAge := int(patient["age"].(float64))
	patientIsPostSurgery := patient["is_post_surgery"].(bool)
	patientSpecialRequirements := patient["special_requirements"].([]interface{})

	result := map[string]interface{}{
		"patient_id":          patientID,
		"assigned_hospital":   nil,
		"assigned_ward":       nil,
		"assigned_bed":        nil,
		"reallocated_patients": []interface{}{},
	}

	var suitableHospitals []map[string]interface{}
	if patientIsPostSurgery {
		for _, h := range hospitals {
			hospital := h.(map[string]interface{})
			if hospital["model"].(int) == 4 {
				suitableHospitals = append(suitableHospitals, hospital)
			}
		}
	} else {
		for _, h := range hospitals {
			hospital := h.(map[string]interface{})
			suitableHospitals = append(suitableHospitals, hospital)
		}
	}

	// Direct Assignment Attempt
	for _, h := range suitableHospitals {
		hospital := h.(map[string]interface{})
		hospitalID := hospital["id"].(string)
		wards := hospital["wards"].([]interface{})

		for _, w := range wards {
			ward := w.(map[string]interface{})
			wardName := ward["ward_name"].(string)
			genderRestriction := ward["gender_restriction"].(string)
			ageRestriction := ward["age_restriction"].([]interface{})
			specialRequirements := ward["special_requirements"].([]interface{})
			availableBeds := ward["available_beds"].([]interface{})

			ageMin := int(ageRestriction[0].(float64))
			ageMax := int(ageRestriction[1].(float64))

			genderMatches := (genderRestriction == "No Restriction" || genderRestriction == patientGender)
			ageMatches := (patientAge >= ageMin && patientAge <= ageMax)

			specialRequirementsMatch := true
			for _, req := range patientSpecialRequirements {
				found := false
				for _, wardReq := range specialRequirements {
					if req == wardReq {
						found = true
						break
					}
				}
				if !found {
					specialRequirementsMatch = false
					break
				}
			}

			if genderMatches && ageMatches && specialRequirementsMatch {
				sort.Slice(availableBeds, func(i, j int) bool {
					return availableBeds[i].(float64) < availableBeds[j].(float64)
				})

				if len(availableBeds) > 0 {
					bedNumber := int(availableBeds[0].(float64))
					result["assigned_hospital"] = hospitalID
					result["assigned_ward"] = wardName
					result["assigned_bed"] = bedNumber

					newAvailableBeds := []interface{}{}
					for _, bed := range availableBeds {
						if int(bed.(float64)) != bedNumber {
							newAvailableBeds = append(newAvailableBeds, bed)
						}
					}
					ward["available_beds"] = newAvailableBeds

					return result
				}
			}
		}
	}

	// Reallocation Attempt (only for high-risk patients)
	if patientIsPostSurgery {
		for _, h := range suitableHospitals {
			hospital := h.(map[string]interface{})
			hospitalID := hospital["id"].(string)
			wards := hospital["wards"].([]interface{})

			for _, w := range wards {
				ward := w.(map[string]interface{})
				wardName := ward["ward_name"].(string)
				currentPatients := ward["current_patients"].([]interface{})

				for _, p := range currentPatients {
					existingPatient := p.(map[string]interface{})
					existingPatientID := existingPatient["id"].(string)
					existingPatientGender := existingPatient["gender"].(string)
					existingPatientAge := int(existingPatient["age"].(float64))
					existingPatientDaysInHospital := int(existingPatient["days_in_hospital"].(float64))
					existingPatientBedNumber := int(existingPatient["bed_number"].(float64))
					existingPatientNonTransferable := existingPatient["non_transferable"].(bool)

					if existingPatientDaysInHospital > 3 && !existingPatientNonTransferable {
						// Find a lower-level hospital for the existing patient
						for _, lowerH := range hospitals {
							lowerHospital := lowerH.(map[string]interface{})
							lowerHospitalID := lowerHospital["id"].(string)
							lowerHospitalModel := lowerHospital["model"].(int)

							if hospital["model"].(int) > lowerHospitalModel {
								lowerHospitalWards := lowerHospital["wards"].([]interface{})

								for _, lowerW := range lowerHospitalWards {
									lowerWard := lowerW.(map[string]interface{})
									lowerWardName := lowerWard["ward_name"].(string)
									lowerWardGenderRestriction := lowerWard["gender_restriction"].(string)
									lowerWardAgeRestriction := lowerWard["age_restriction"].([]interface{})
									lowerWardSpecialRequirements := lowerWard["special_requirements"].([]interface{})
									lowerWardAvailableBeds := lowerWard["available_beds"].([]interface{})

									lowerAgeMin := int(lowerWardAgeRestriction[0].(float64))
									lowerAgeMax := int(lowerWardAgeRestriction[1].(float64))

									lowerGenderMatches := (lowerWardGenderRestriction == "No Restriction" || lowerWardGenderRestriction == existingPatientGender)
									lowerAgeMatches := (existingPatientAge >= lowerAgeMin && existingPatientAge <= lowerAgeMax)

									lowerSpecialRequirementsMatch := true
									// Check existing patient's special requirements against the new ward
									// Note that we can't access the special_requirements field of the existing patient, so we skip this check
									// for _, req := range existingPatient["special_requirements"].([]interface{}) {
									// 	found := false
									// 	for _, wardReq := range lowerWardSpecialRequirements {
									// 		if req == wardReq {
									// 			found = true
									// 			break
									// 		}
									// 	}
									// 	if !found {
									// 		lowerSpecialRequirementsMatch = false
									// 		break
									// 	}
									// }

									if lowerGenderMatches && lowerAgeMatches && lowerSpecialRequirementsMatch && len(lowerWardAvailableBeds) > 0 {
										// Transfer is possible
										sort.Slice(lowerWardAvailableBeds, func(i, j int) bool {
											return lowerWardAvailableBeds[i].(float64) < lowerWardAvailableBeds[j].(float64)
										})

										newBedNumber := int(lowerWardAvailableBeds[0].(float64))

										// Update available beds in the lower ward
										newLowerWardAvailableBeds := []interface{}{}
										for _, bed := range lowerWardAvailableBeds {
											if int(bed.(float64)) != newBedNumber {
												newLowerWardAvailableBeds = append(newLowerWardAvailableBeds, bed)
											}
										}
										lowerWard["available_beds"] = newLowerWardAvailableBeds

										// Assign the high-risk patient
										result["assigned_hospital"] = hospitalID
										result["assigned_ward"] = wardName
										result["assigned_bed"] = existingPatientBedNumber

										// Record the reallocation
										reallocatedPatient := map[string]interface{}{
											"patient_id":   existingPatientID,
											"new_hospital": lowerHospitalID,
											"new_ward":     lowerWardName,
											"new_bed":      newBedNumber,
										}

										result["reallocated_patients"] = append(result["reallocated_patients"].([]interface{}), reallocatedPatient)
										return result
									}
								}
							}
						}
					}
				}
			}
		}
	}

	return result
}

func main() {
	var input map[string]interface{}
	err := json.NewDecoder(json.NewDecoder(fmt.NewReader(`{"patient": {"id": "P002", "gender": "Male", "age": 50, "condition": "Post-Surgery Recovery", "special_requirements": ["Intensive Monitoring"], "is_post_surgery": true}, "hospitals": [{"id": "H1", "model": 4, "wards": [{"ward_name": "Surgical Recovery Ward", "gender_restriction": "No Restriction", "age_restriction": [18, 80], "special_requirements": ["Intensive Monitoring"], "available_beds": [], "current_patients": [{"id": "P101", "gender": "Female", "age": 65, "condition": "Post-Surgery Recovery", "days_in_hospital": 5, "bed_number": 3, "non_transferable": false}]}]}, {"id": "H3", "model": 2, "wards": [{"ward_name": "General Ward", "gender_restriction": "Female Only", "age_restriction": [18, 85], "special_requirements": [], "available_beds": [1, 2, 3], "current_patients": []}]}]}`)).Decode(&input)).Decode(&input)
	if err != nil {
		panic(err)
	}

	patient := input["patient"].(map[string]interface{})
	hospitals := input["hospitals"].([]interface{})

	result := reallocateBed(patient, hospitals)

	jsonResult, err := json.Marshal(result)
	if err != nil {
		panic(err)
	}

	fmt.Println(string(jsonResult))
}

File: codes/python/gemini/minimal/task_2/Solution.py
import json

def reallocate_bed(patient, hospitals):
    patient_id = patient.get('id')
    patient_gender = patient.get('gender')
    patient_age = patient.get('age')
    patient_is_post_surgery = patient.get('is_post_surgery')
    patient_special_requirements = patient.get('special_requirements')

    assigned_hospital = None
    assigned_ward = None
    assigned_bed = None
    reallocated_patients = []

    def find_matching_wards(hospital, patient):
        matching_wards = []
        for ward in hospital['wards']:
            gender_restriction = ward.get('gender_restriction')
            age_restriction = ward.get('age_restriction')
            special_requirements = ward.get('special_requirements')

            gender_match = (gender_restriction == "No Restriction" or gender_restriction == patient['gender'] + " Only")
            age_match = (age_restriction[0] <= patient['age'] <= age_restriction[1])

            special_requirements_match = all(req in special_requirements for req in patient['special_requirements'])

            if gender_match and age_match and special_requirements_match:
                matching_wards.append(ward)
        return matching_wards

    def direct_assignment(patient, hospitals):
        for hospital in hospitals:
            if patient['is_post_surgery'] and hospital['model'] != 4:
                continue
            if not patient['is_post_surgery'] and hospital['model'] == 4:
                continue

            matching_wards = find_matching_wards(hospital, patient)

            for ward in matching_wards:
                available_beds = sorted(ward.get('available_beds'))
                for bed in available_beds:
                    return hospital['id'], ward['ward_name'], bed
        return None, None, None

    def reallocation(patient, hospitals):
        if not patient['is_post_surgery']:
            return None, []

        for hospital in hospitals:
            if hospital['model'] != 4:
                continue

            for ward in hospital['wards']:
                current_patients = ward.get('current_patients')
                if not current_patients:
                    continue

                for existing_patient in sorted(current_patients, key=lambda x: x.get('bed_number')):
                    if existing_patient['days_in_hospital'] > 3 and not existing_patient['non_transferable']:
                        transfer_patient_id = existing_patient['id']
                        transfer_bed_number = existing_patient['bed_number']

                        for lower_hospital in hospitals:
                            if lower_hospital['model'] >= 4:
                                continue

                            lower_hospital_wards = find_matching_wards(lower_hospital, existing_patient)

                            if not lower_hospital_wards:
                                continue

                            for lower_ward in lower_hospital_wards:
                                available_beds = sorted(lower_ward.get('available_beds'))
                                if available_beds:
                                    new_bed = available_beds[0]
                                    reallocated_patients.append({
                                        'patient_id': transfer_patient_id,
                                        'new_hospital': lower_hospital['id'],
                                        'new_ward': lower_ward['ward_name'],
                                        'new_bed': new_bed
                                    })
                                    return hospital['id'], ward['ward_name'], transfer_bed_number

        return None, []

    assigned_hospital, assigned_ward, assigned_bed = direct_assignment(patient, hospitals)

    if assigned_hospital:
        return {
            'patient_id': patient_id,
            'assigned_hospital': assigned_hospital,
            'assigned_ward': assigned_ward,
            'assigned_bed': assigned_bed,
            'reallocated_patients': []
        }

    assigned_hospital, ward_name, assigned_bed = reallocation(patient, hospitals)
    if assigned_hospital:
        return {
            'patient_id': patient_id,
            'assigned_hospital': assigned_hospital,
            'assigned_ward': ward_name,
            'assigned_bed': assigned_bed,
            'reallocated_patients': reallocated_patients
        }

    return {
        'patient_id': patient_id,
        'assigned_hospital': None,
        'assigned_ward': None,
        'assigned_bed': None,
        'reallocated_patients': []
    }

if __name__ == "__main__":
    import sys

    input_data = json.loads(sys.stdin.read())
    patient = input_data.get('patient')
    hospitals = input_data.get('hospitals')

    result = reallocate_bed(patient, hospitals)
    print(json.dumps(result))